### PR TITLE
fix cases where optimizing out dropping/creating the same index in on…

### PR DESCRIPTION
**Pull request by alxyang**
…e DDL statement improperly removed the index from data dictionary

Fixes https://github.com/facebook/mysql-5.6/issues/602.

### CODE REVIEW

**Comment by alxyang** [comment]
actually, this isn't ready for review yet. there are some corner cases I need to cover, and need to add tests.

File: storage/rocksdb/ha_rocksdb.cc [hunk 1]
```diff
@@ -5422,9 +5423,32 @@ std::unordered_map<std::string, uint> ha_rocksdb::get_old_key_positions(
...
-    if (!compare_key_parts(old_key, new_key)) {
-      old_key_pos[old_key->name] = i;
+    /* If index algorithms are different we need to rebuild. */
+    if (old_key->algorithm != new_key->algorithm) {
+      continue;
+    }
+
+    /*
+      Check that the key is identical between old and new tables.
+      If so, we still need to create a new index.
```
Comment (hermanlee) [inline storage/rocksdb/ha_rocksdb.cc]: Should say "If not, we still need to create a new index."

File: storage/rocksdb/ha_rocksdb.cc [hunk 2]
```diff
@@ -5422,9 +5423,32 @@ std::unordered_map<std::string, uint> ha_rocksdb::get_old_key_positions(
...
+
+      The exception is if there is an index changed from unique to non-unique,
+      in these cases we don't need to rebuild as they are stored the same way in
+      RocksDB.
+    */
+    bool unique_to_non_unique =
+        (old_key->flags ^ new_key->flags) == HA_NOSAME &&
+        (old_key->flags & HA_NOSAME);
+    if (!unique_to_non_unique && ((old_key->flags & HA_KEYFLAG_MASK) !=
```
Comment (hermanlee) [inline storage/rocksdb/ha_rocksdb.cc]: If the mask is needed, then the mask should apply to the old_key->flags & new_key->flags calculation. Logically, your code works too though.

**Review by hermanlee — COMMENTED** [review]
What does InnoDB do in the case where the same key is dropped and added? Does it skip over it, or does it drop the index and re-create it, even though it is the same?

**Comment by alxyang** [comment]
@hermanlee if an index is exactly the same (meaning all the key flags match) and the drop key + add key is combined in one alter statement, then it's actually optimized out at the SQL layer.  So yes, it skips over it (does not recreate).
